channel2/wss_listener.go

```diff
@@ -0,0 +1,280 @@
+
+	if err != nil {
+		log.WithField("err", err).Error("websocket upgrade failed. Failure not recoverable. Exiting listen loop")
+	} else {
+
+		socket := c.UnderlyingConn() // Obtain the socket underneath the websocket
+
+		connection := &WssConnection{
+			detail: &transport.ConnectionDetail{
+				Address: "wss:" + socket.RemoteAddr().String(),
+				InBound: true,
+				Name:    "wss",
+			},
+			websocket: c,
+			socket:    socket.(*tls.Conn),
+		}
+
+		listener.incoming <- connection // pass the socket to the goroutine that will validate the HELLO handshake
+	}
+}
+
+/**
+ *	Simply start the goroutines that will process incoming websocket connections
+ */
+func (listener *wssListener) Listen(handlers ...ConnectionHandler) error {
+	go listener.wsslistener()
+	go listener.binaryListener()
+	return nil
+}
+
+/**
+ *	The TLS-based listener that accepts incoming HTTP connections that we need to upgrade to websocket connections
+ */
+func (listener *wssListener) wsslistener() {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("starting HTTP (websocket) server")
+
+	router := mux.NewRouter()
+
+	router.HandleFunc("/wss", listener.handleWebsocket).Methods("GET")
+
+	httpServer := &http.Server{
+		Addr:         listener.endpoint.BindableAddress(),
+		WriteTimeout: time.Second * 15,
+		ReadTimeout:  time.Second * 15,
+		IdleTimeout:  time.Second * 60,
+		Handler:      router,
+		TLSConfig: &tls.Config{
+			ClientAuth: tls.RequestClientCert,
+		},
+	}
+
+	// log.Info("listener.options: [%v]", listener.options)
+
+	if err := httpServer.ListenAndServeTLS(listener.serverCert, listener.key); err != nil {
+		panic(err)
+	}
+}
+
+func (listener *wssListener) Close() error {
+	close(listener.close)
+	close(listener.created)
+	if err := listener.socket.Close(); err != nil {
+		return err
+	}
+	listener.socket = nil
+	return nil
+}
+
+func (listener *wssListener) Create(tcfg transport.Configuration) (Underlay, error) {
+	if listener.created == nil {
+		return nil, errors.New("closed")
+	}
+	impl := <-listener.created
+	if impl == nil {
+		return nil, errors.New("closed")
+	}
+	return impl, nil
+}
+
+/**
+ *	Take the incoming (now binary) websocket
+ */
+func (listener *wssListener) binaryListener() {
+	log := pfxlog.ContextLogger(listener.endpoint.String())
+	log.Info("started")
+	defer log.Info("exited")
+
+	for {
+		select {
+		case peer := <-listener.incoming:
+
+			impl := newWssImpl(peer, 2)
+			if connectionId, err := globalRegistry.newConnectionId(); err == nil {
+				log.Infof("hello: new connectionId: %v", connectionId)
+
+				impl.connectionId = connectionId
+				request, hello, err := listener.receiveHello(impl)
+				if err == nil {
+					for _, h := range listener.handlers {
+						log.Infof("hello: %v, peer: %v, handler: %v", hello, peer, h)
+						if err := h.HandleConnection(hello, peer.PeerCertificates()); err != nil {
+							log.Errorf("connection handler error (%s)", err)
```

Comment:
When wrapping errors `%v` should be used.

`...Errorf("some error happened because of this error: %v", err)`

Reply:
fixed